### PR TITLE
internal/charmstore: improve search results for charm names

### LINKED ISSUE #638: Search shouldn't be case sensitive
See the two queries:

lowercase does return the apache2 promulgated charm.
https://api.jujugui.org/charmstore/v4/search?text=apache2&tags&limit=150&include=charm-metadata&include=charm-config&include=bundle-metadata&include=extra-info&include=tags&include=stats

Uppercase doesn't return the apache2 promulgated charm.
https://api.jujugui.org/charmstore/v4/search?text=APACHE2&tags&limit=150&include=charm-metadata&include=charm-config&include=bundle-metadata&include=extra-info&include=tags&include=stats

### PR DESCRIPTION
**Pull request by mhilton**
Update the elasticsearch indexing so that search is case insensitive
and name matching matches on tokens.

This fixes #638 & #639.

### CODE REVIEW

**Comment by mhilton** [comment]
## QA

```
$ curl "localhost:8080/v5/search?text=sql&type=charm&limit=100&autocomplete=1&promulgated=1"
{"SearchTime":32000000,"Total":7,"Results":[{"Id":"cs:trusty/postgresql-psql-1"},{"Id":"cs:precise/postgresql-psql-7"},{"Id":"cs:trusty/mysqltest-0"},{"Id":"cs:trusty/mysql-36"},{"Id":"cs:trusty/postgresql-41"},{"Id":"cs:precise/mysql-51"},{"Id":"cs:precise/postgresql-98"}]}
$ curl "localhost:8080/v5/search?text=SQL&type=charm&limit=100&autocomplete=1&promulgated=1"
{"SearchTime":6000000,"Total":7,"Results":[{"Id":"cs:trusty/postgresql-psql-1"},{"Id":"cs:precise/postgresql-psql-7"},{"Id":"cs:trusty/mysqltest-0"},{"Id":"cs:trusty/mysql-36"},{"Id":"cs:trusty/postgresql-41"},{"Id":"cs:precise/mysql-51"},{"Id":"cs:precise/postgresql-98"}]}
$ curl "localhost:8080/v5/search?text=opensta&type=bundle&limit=100&autocomplete=1&promulgated=1"
{"SearchTime":2000000,"Total":2,"Results":[{"Id":"cs:bundle/openstack-telemetry-33"},{"Id":"cs:bundle/openstack-base-40"}]}
```

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1812/
Test PASSed.

File: internal/charmstore/search_test.go [hunk 1]
```diff
@@ -657,6 +667,24 @@ var searchTests = []struct {
...
+		},
+	}, {
+		about: "case insensitive search",
+		sp: SearchParams{
+			Text: "WORDPRESS",
+		},
+		results: Entities{
+			searchEntities["wordpress"].entity,
+			searchEntities["wordpress-simple"].entity,
+		},
```
Comment (rogpeppe) [inline internal/charmstore/search_test.go:697]: Wasn't one of the failing cases where someone searching for "mongo" didn't find "mongodb"?

**Comment by rogpeppe** [comment]
I'd like to see some of the cases from the QA suggestions in the tests, but
LGTM otherwise, assuming QA ok.

File: internal/charmstore/search_test.go (hunk 1 above)
Comment (mhilton) [inline internal/charmstore/search_test.go:697]: It was that is covered by a later test.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1813/
Test PASSed.

**Comment by urosj** [comment]
👍 QA 🆗

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**